src/Algebra/Graph/AdjacencyIntMap/Algorithm.hs

```diff
@@ -169,10 +184,27 @@ dfsForest = Typed.dfsForest . Typed.fromAdjacencyIntMap
 -- 'forest' (dfsForestFrom [3] $ 'circuit' [1..5] + 'circuit' [5,4..1]) == 'path' [3,2,1,5,4]
 -- @
 dfsForestFrom :: [Int] -> AdjacencyIntMap -> Forest Int
-dfsForestFrom vs = Typed.dfsForestFrom vs . Typed.fromAdjacencyIntMap
+dfsForestFrom vs g = dfsForestFrom' [ v | v <- vs, hasVertex v g ] g
+
+dfsForestFrom' :: [Int] -> AdjacencyIntMap -> Forest Int
+dfsForestFrom' vs g = evalState (explore vs) IntSet.empty where
+  explore (v:vs) = discovered v >>= \case
```

Comment:
Isn't `explore` just a `foldr` under the hood ?

Reply:
We want it to be foldl to expand the from the smallest . Also, if a monadielementsc foldl is used, the components would come back reversed, so I went with explicit recursion. I may be missing a nicer way to express this of course

Reply:
I guess like a `filterMapM` type thing is what's wanted, of a `mapM f`, where `f :: a -> Maybe (Tree a)`, but that adds unnecessary indirection

Reply:
The way I really wanted to express it was using 

`unfoldForestM :: Monad m => (b -> m (a, [b])) -> [b] -> m (Forest a)` 

The issue was that each `b` is given a tree, so I ended up to wrapping them with `Maybe` and pruning the result to discard the `Nothing` nodes.

Reply:
I don't have a strong objection to using explicit recursion here but this does look like `foldlM` to me:

https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Foldable.html#v:foldlM

@jitwit Could you clarify why `foldlM` doesn't work?

Reply:
It would work in the sense that smaller vertices get explored first, but the forest comes out reversed:
```
λ> map flatten $ dfsForest $ path [1..3] + path [4..6]
[[4,5,6],[1,2,3]]
```
That's with internal functions written as:
```
search = foldlM explore' [] vs
explore' vs v = discovered v >>= \case
  True -> (:vs) <$> walk v
  False -> return vs
```

Reply:
> but the forest comes out reversed

Well, this is expected since you reverse the result by placing `vs` at the end in `(:vs) <$> walk v`.

I see what you mean though: DFS is more naturally expressed using your custom function `explore`, so let's keep it :)